test/lib/mock-blpapi.ts

```diff
@@ -0,0 +1,272 @@
+    }
+
+    authorizeUser(request: any, cid: number): number {
+        throw new Error('Not yet implemented');
+    }
+
+    stop(): Session {
+        throw new Error('Not yet implemented');
+    }
+
+    // Don't see a test case for this function yet
+    destroy(): Session {
+        return this;
+    }
+
+    openService(uri: string, cid: number): number {
+        // If has preset instructions, follow it
+        if (_.has(instructions, 'openService')) {
+            process.nextTick((): void => {
+                if (instructions.openService) {
+                    this.sendServiceOpened(cid);
+                } else {
+                    this.sendServiceOpenFailure(uri, cid);
+                }
+            });
+        } else {    // Otherwise, wait for instruction
+            ipc.emit('wait-to-openService', {
+                'uri': uri,
+                'cid': cid
+            });
+            ipc.on(util.format('openService-%d-success', cid), (): void => {
+                this.sendServiceOpened(cid);
+            });
+            ipc.on(util.format('openService-%d-fail', cid), (): void => {
+                this.sendServiceOpenFailure(uri, cid);
+            });
+        }
+        return cid;
+    }
+
+    subscribe(subs: Subscription[], identity?: Identity, label?: string): Session {
+        // For subscribe, only communicate via ipc is supported
+        ipc.emit('wait-to-subscribe', subs);
+        _.forEach(subs, (sub: Subscription): void => {
+            var serviceUri = getServiceForSecurity(sub.security);
+            assert(SERVICE_TO_SUBSCRIPTION_EVENTS_MAP[serviceUri],
+                   'Invalid service name.');
+            var events: string[] = SERVICE_TO_SUBSCRIPTION_EVENTS_MAP[serviceUri];
+            _.forEach(events, (e: string): void => {
+                ipc.on(util.format('subscription-%d-%s', sub.correlation, e), (): void => {
+                    this.sendSubscriptionData(e, sub.correlation);
+                });
+            });
+        });
+        return this;
+    }
+
+    resubscribe(subs: Subscription[], label?: string): Session {
+        throw new Error('Not yet implemented');
+    }
+    unsubscribe(subs: Subscription[]): Session {
+        // As long as we no longer send data event via ipc for unsubscribed cids, no-ops
+        return this;
+    }
+    request(uri: string,
+            name: string,
+            request: any,
+            cid: number,
+            identity?: Identity,
+            label?: string): number {
+        assert(REQUEST_TO_RESPONSE_MAP[name], 'Invalid request name.');
+        var responseName = REQUEST_TO_RESPONSE_MAP[name];
+        // If has preset instructions, follow it
+        if (_.has(instructions, 'request')) {
+            // throw immediately if the first instruction is throwError
+            if ('throwError' === instructions.request[0]) {
+                this.throwError();
+            } else {
+                Promise.resolve(instructions.request)
```

Comment:
`Promise.resolve().each()` I think can be replaced with `Promise.each(iterable, function)`.